File: src/main/kotlin/org/gameflow/Game.kt
package org.gameflow

import java.util.List
import java.util.Map
import java.util.HashMap
import org.gameflow.state.GameState
import org.gameflow.util.ListenableCollection
import java.util.concurrent.ConcurrentHashMap
import org.gameflow.state.StateManager
import org.gameflow.state.StateManagerImpl
import org.gameflow.entity.EntityGroup
import org.gameflow.entity.EntityGroupsImpl
import org.gameflow.entity.EntityGroups
import java.util.ArrayList
import org.gameflow.pass.Pass
import java.util.Collection
import org.gameflow.pass.Passable
import java.util.Collections
import org.gameflow.entity.BaseEntity
import org.gameflow.clock.ClockImpl
import java.util.concurrent.ConcurrentSkipListSet

/**
 * 
 */
public open class Game(
        public val clock: Clock = ClockImpl(),
        public val stateManager: StateManager = StateManagerImpl(),
        public val groups : EntityGroups = EntityGroupsImpl()
    ):
        BaseEntity(),
        Clock by clock,
        EntityGroups by groups,
        StateManager by stateManager  {

    private val passes : Collection<Pass> = ConcurrentSkipListSet<Pass>()


    public open fun main(args: Array<String>) {
        start()
    }

    public fun addPass(pass: Pass) {
        if (!passes.contains(pass)) {
            passes.add(pass)
            if (started) pass.init(this)
        }
    }

    public fun removePass(pass: Pass) {
        if (passes.contains(pass)) {
            passes.remove(pass)
            if (started) pass.deInit(this)
        }
    }

    public fun hasPass(pass: Pass): Boolean = passes.contains(pass)
    public fun getPasses(): Collection<Pass> = Collections.unmodifiableCollection(passes) !!

    private var started = false
    private var stopped = false

    /** Calls init, then starts the game loop.  Call stop to exit it. */
    public final fun start() {
        if (!started) {
            started = true

            for (pass in passes) { pass.init(this) }
            init()

            while(!stopped) {
                doLoop()
            }

            deInit()
            for (pass in passes) { pass.deInit(this) }
        }
    }

    /** Causes the game to stop after the current run through the main loop finishes */
    public final fun stop() {
        stopped = true
    }

    /** Runs through the game loop once. */
    public final fun doLoop() {
        // Update time
        clock.tick()

        // Change game state if needed
        stateManager.doGameStateUpdate(this)

        // Run passes in order
        for (pass in passes) {
            pass.startPass(this)

            // Apply pass to game and all sub-components
            doPass(pass, this)

            pass.endPass(this)
        }
    }

    override fun doPassOnChildren(pass : Pass, game: Game) {
        groups.doPass(pass, game)
        stateManager.doPass(pass, game)
    }

    /** Called when the game loop is started. */
    public open fun init() {}

    /** Called after the game loop is stopped. */
    public open fun deInit() {}

}

File: src/main/scala/org/gameflow/core/Game.scala
package org.gameflow.core

import clock.{FixedDelayClock, Clock}
import entity.EntityGroupsImpl
import pass.{PassManager, Pass, Passable}
import scala.collection.JavaConversions._
import state.GameStateManager
import org.gameflow.utils.ParameterChecker
import java.util.{Collections, ArrayList}

/**
 * Root game class.  Does not have any managers pre-added, usually you want to use a more specific derived class of this.
 */
class Game extends Passable {

  private val _managers = new java.util.ArrayList[Manager]()

  private var initialized = false
  private var stopped = false
  private var running = false


  // Lifecycle state
  final def isRunning = running
  final def isInitialized = initialized
  final def isStopped = stopped


  // Main method

  /**
   * Default main method that calls start, provided for convenience in case this class is overridden by an object.
   */
  def main(args: Array[String]) {
    start()
  }


  // Main loop

  /**
   * Calls init, then starts the game loop.  Call stop to exit it.
   * A game can only be started once.
   * Calls de-init when loop ends.
   */
  final def start() {
    init()
    runLoop()
    deInit()
  }

  /**
   * Causes the game to stop after the current run through the main loop finishes
   */
  final def stop() {
    stopped = true
  }



  // Managers

  final def addManager[T <: Manager](manager: T): T = {
    _managers.add(manager)

    if (initialized) manager.init(this)

    manager
  }

  final def managers: java.util.List[Manager] = Collections.unmodifiableList(_managers)


  // Passable

  override final def doPassOnChildren(pass: Pass, game: Game) {
    _managers.foreach(_.doPass(pass, game))
  }


  // Over-rideable init and update

  /** Called before managers are initialized, can be used for adding any necessary managers or doing other early setup work. */
  protected def setup() {}

  /** Called after managers have been created and initialized, can be used for setting up game content. */
  protected def onInit() {}

  /** Called each update tick. */
  protected def onUpdate() {}

  /** Called when main loop has been exited and application is about to quit, but before managers have been de-initialized. */
  protected def onDeInit() {}


  // Game lifecycle and loop parts

  protected final def init() {
    if (initialized) throw new Error("Game '" + this + "' has already been initialized.")

    setup()
    _managers.foreach(_.init(this))
    onInit()

    initialized = true
  }

  protected final def runLoop() {
    if (!initialized) throw new Error("Game '" + this + "' has not yet been initialized.")

    running = true
    while (!stopped) {
      loopOnce()
    }
    running = false
  }

  /** Runs through the game loop once. */
  protected final def loopOnce() {
    if (!initialized) throw new Error("Game '" + this + "' has not yet been initialized.")

    _managers.foreach(_.update(this))

    onUpdate()
  }

  protected final def deInit() {
    if (!initialized) throw new Error("Game '" + this + "' has not yet been initialized.")
    if (!stopped) throw new Error("Game '" + this + "' has not yet been stopped.")

    onDeInit()
    _managers.toList.reverse.foreach(_.deInit(this)) // De-initialize in reverse order of initialization.
  }


}
